Gems/Atom/RPI/Code/Source/RPI.Builders/Material/MaterialBuilder.cpp

```diff
@@ -117,7 +117,7 @@ namespace AZ
                 // We usually won't load file during CreateJob since we want to keep the function fast. But here we have to load the
                 // material type data to find the exact material type format so we could create an accurate source dependency.
                 const auto& resolvedMaterialTypePath =
-                    AssetUtils::ResolvePathReference(materialSourcePath, materialSourceData.m_materialType);
+                    MaterialUtils::PredictOriginalMaterialTypeSourcePath(AssetUtils::ResolvePathReference(materialSourcePath, materialSourceData.m_materialType));
```

Comment:
I assume this new function was already present somewhere else?

Reply:
I'm not sure this is the correct fix. If the original material type is abstract, the material should end up pointing to the material type in the intermediate asset folder and failing until it is generated on a subsequent AP pass. 

The way the dependencies are set up should cause the material to reprocess after all of the intermediate assets have been processed and resolved. That is the part that is failing in the AP batch, but only for some materials, possibly the last round processed. 

Do all of the materials load correctly after being processed with these changes? There should be an assert somewhere if attempting to load a material asset that's referencing an abstract material type.

Reply:
The source material files point to the original material Como which has two formats, direct or abstract.

A direct material type is concrete with all of the shaders, functors, and properties filled in. It can be referenced and used directly to instantiate materials.

An abstract material type is incomplete, specifies shader template references, functors, and lighting models, for the builder to transform into a fully generated, direct material type gets saved into the intermediate assets folder.

The material type class has a function that evaluates the contents to determine if its format is direct or abstract.

By always resolving to the original material type path, you will always be loading the abstract version if that was the original format. That changes the logic and dependencies of the code below your change. I think this will also break hot reloading and tools like material canvas that will be notified prematurely that assets are complete, because this will ignore dependencies on the direct intermediate materials and their nested dependencies. Please confirm this.

The fail until all dependencies are available for the asset to be successfully processed logic is the nature of the intermediate asset system, not materials. It should be silent in the GUI AP, and always resolve unless there was another problem.

My understanding of the problem this is trying to address is that it is specific to the AP batch. I suspect that there is some difference in code paths between AP batch and the regular AP. I should have some time to help investigate this.

Reply:
What's broken in the AP GUI? I have not heard any reports of those assets not resolving after all dependency passes have completed.

Open material canvas.
Open or create a material graph.
Manipulate the material graph by creating nodes, editing node properties, using the property editor, deleting nodes, changing connections, etc.

Confirm that all references to the material update and reflect the state of the material graph after making changes to the graph, and the resultant material types, materials, and shaders have finished compiling in the AP.

This includes the material canvas viewport, asset browser thumbnails, asset browser previews, material component previews, material instance editor previews, material editor viewport, main editor viewport, etc.

They're supposed to update correctly after hand editing shader files used in those material types.

Reply:
I see now…

Source materials should not point to one directly in the intermediate assets folder. That is a problem that would only happen if someone was doing it by hand, created it from the main editor asset browser directly from an intermediate asset, or they created it using "Generate Source Materials" right after intermediate assets were introduced.

The asset system does not track the connection or dependency between the intermediate asset and the asset that created it. So "Generate Source Materials" caused generated materials to reference the intermediate asset material type. That is actually why I added the function you're calling in the PR. I get it now that I see the assets.

Part of me says to make this an error case, fail the asset, and inform the user on how to fix it.

But if the above code works to automatically resolve the problem and let the assets get processed then I would at least detect that it's referencing an intermediate asset first and make it a warning to notify the user, even though it's being fixed up.